snuba/consumer.py

```diff
@@ -356,3 +362,226 @@ def create(
+            if timeout is not None:
+                timeout_remaining: Optional[float] = max(
+                    timeout - (time.time() - start), 0
+                )
+            else:
+                timeout_remaining = None
+
+            step.join(timeout_remaining)
+
+
+def process_message_multistorage(
+    message: Message[Tuple[Sequence[StorageKey], KafkaPayload]]
+) -> Sequence[Tuple[StorageKey, Union[None, JSONRowInsertBatch, ReplacementBatch]]]:
+    # XXX: Avoid circular import on KafkaMessageMetadata, remove when that type
+    # is itself removed.
+    from snuba.datasets.storages.factory import get_writable_storage
+
+    storage_keys, payload = message.payload
+    value = rapidjson.loads(payload.value)
+    metadata = KafkaMessageMetadata(
+        message.offset, message.partition.index, message.timestamp
+    )
+
+    results: MutableSequence[
+        Tuple[StorageKey, Union[None, JSONRowInsertBatch, ReplacementBatch]]
+    ] = []
+
+    for storage_key in storage_keys:
+        result = (
+            get_writable_storage(storage_key)
+            .get_table_writer()
+            .get_stream_loader()
+            .get_processor()
+            .process_message(value, metadata)
+        )
+        if isinstance(result, InsertBatch):
+            results.append(
+                (
+                    storage_key,
+                    JSONRowInsertBatch(
+                        [json_row_encoder.encode(row) for row in result.rows]
+                    ),
+                )
+            )
+        else:
+            results.append((storage_key, result))
+
+    return results
+
+
+class MultistorageConsumerProcessingStrategyFactory(
+    ProcessingStrategyFactory[KafkaPayload]
+):
+    def __init__(
+        self,
+        storages: Sequence[WritableTableStorage],
+        max_batch_size: int,
+        max_batch_time: float,
+        processes: int,
+        input_block_size: int,
+        output_block_size: int,
+        metrics: MetricsBackend,
+    ):
+        self.__storages = storages
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+        self.__processes = processes
+        self.__input_block_size = input_block_size
+        self.__output_block_size = output_block_size
+        self.__metrics = metrics
+
+    def __find_destination_storages(
+        self, message: Message[KafkaPayload]
+    ) -> Tuple[Sequence[StorageKey], KafkaPayload]:
+        storages: MutableSequence[StorageKey] = []
+        for storage in self.__storages:
+            filter = storage.get_table_writer().get_stream_loader().get_pre_filter()
+            if filter is None or not filter.should_drop(message):
+                storages.append(storage.get_storage_key())
+        return (storages, message.payload)
+
+    def __has_destination_storages(
+        self, message: Message[Tuple[Sequence[StorageKey], KafkaPayload]]
+    ) -> bool:
+        return len(message.payload[0]) > 0
```

Comment:
This seems like it would always be true now. Should this be the length of prefiltered storages, or is that handled at another step now?

Reply:
Made that change with 0ea85cd, let me know if that is/isn't more readable or if I misinterpreted your comment.

Reply:
I missed that the storages were prefiltered before the FilterStep earlier - this all makes sense. 👍 for NamedTuple.